src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -26,73 +26,61 @@ private void FinalizeDispose()
-                                    Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved |
-                                    Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed;
-                }
+                return;
+            }
+            else
+            {
+                _cancellation = new CancellationTokenSource();
+            }
 
-                CreateStreamAndStartWatcher();
+            Interop.EventStream.FSEventStreamEventFlags flags = 0;
+
+            // Always re-create the filter flags when start is called since they could have changed
+            if ((_notifyFilters & (NotifyFilters.Attributes | NotifyFilters.CreationTime | NotifyFilters.LastAccess | NotifyFilters.LastWrite | NotifyFilters.Size)) != 0)
+            {
+                flags = Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemInodeMetaMod  | 
+                               Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemFinderInfoMod |
+                               Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemModified      |
+                               Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemChangeOwner;
             }
+            if ((_notifyFilters & NotifyFilters.Security) != 0)
+            {
+                flags |= Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemChangeOwner | Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemXattrMod;
+            }
+            if ((_notifyFilters & NotifyFilters.DirectoryName) != 0)
+            {
+                flags |= Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsDir |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsSymlink |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed;
+            }
+            if ((_notifyFilters & NotifyFilters.FileName) != 0)
+            {
+                flags |= Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsFile |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsSymlink |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed;
+            }
```

Comment:
Consider separating all of this NotifyFilters->FSEventStreamEventFlags translation into its own helper static method.